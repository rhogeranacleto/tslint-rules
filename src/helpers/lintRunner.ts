import { Configuration, Linter, LintResult } from 'tslint';

interface IMaria {
	src: string;
	// tslint:disable-next-line:no-any
	rule: any;
	fileName?: string;
	fix?: boolean;
}

export function lintRunner(data: IMaria) {

	const { src, rule, fileName = `/tmp/test-${Date.now()}.snap`, fix = false } = data;

	const linter = new Linter({ fix });

	linter.lint(
		fileName,
		src,
		Configuration.parseConfigFile({
			rules: {
				[rule.name || rule]: [true]
			},
			rulesDirectory: 'src'
		})
	);

	return linter.getResult();
}

export function getReplacements(result: LintResult) {

	if (result.fixes) {

		return result.fixes.map(fix => {

			const fixJson = fix.toJson().fix;

			if (fixJson) {

				if (Array.isArray(fixJson)) {

					return fixJson.map(replace => replace.innerText);
				}

				return [fixJson.innerText];
			}
		});
	}
}