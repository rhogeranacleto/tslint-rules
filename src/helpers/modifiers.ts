export function removeLastEmptyLineOfTheBlock(block: string) {

	const lines = block.split('\n');
	const bracket = lines.pop();

	for (let i = lines.length - 1; i > 0; i--) {

		if (lines[i].trim()) {

			break;
		}

		lines.pop();
	}

	lines.push(bracket as string);

	return lines.join('\n');
}