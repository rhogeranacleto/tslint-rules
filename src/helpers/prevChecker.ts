// tslint:disable-next-line:no-implicit-dependencies
import { getPreviousStatement, getPreviousToken } from 'tsutils';
import ts from 'typescript';

export function prevStatementChecker(node: ts.Statement, sourceFile: ts.SourceFile) {

	const prev = getPreviousStatement(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

export function prevTokenChecker(node: ts.Node, sourceFile: ts.SourceFile) {

	const prev = getPreviousToken(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

function checkPrevLine(node: ts.Node, sourceFile: ts.SourceFile, prev: ts.Node) {

	const startLine = ts.getLineAndCharacterOfPosition(sourceFile, node.getStart()).line;
	const prevLine = ts.getLineAndCharacterOfPosition(sourceFile, prev.getEnd()).line;

	return prevLine === (startLine - 1);
}