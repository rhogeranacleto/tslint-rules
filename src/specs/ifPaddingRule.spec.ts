import { lintRunner } from '../helpers/lintRunner';

const RULE = 'if-padding';

describe('LastLineRule', () => {

	it.each([
		[`console.log(true);
		if(true){
			return true;
		}`, 1],
		[`console.log(true);
		if(true){
			return true;
		}else{
			return false;
		}`, 1],
		[`if(true){
			return 'só if';
		}
		else {
			return 'so else';
		}`, 1],
		[`if(true){
			return 'so if';
		}
		else
		if(false&&maria) {
			return 'else if';
		}
		else {
			return 'só else';
		}`, 2],
		[`if(true){
			return 'so if';
		}
		else if(false&&maria) {
			return 'else if';
		}`, 1],
		[`if(true){
			return 'so if';
		} else
		{
			return 'else if';
		}`, 1],
		[`if(true){
			return 'so if';
		} else if(false&&maria) {
			return 'else if';
		} else {
			return 'só else';
		}`, 0],
		[`console.log(true);

		if(true){
			return true;
		}`, 0]
	])('%s', (src, count) => {

		const result = lintRunner({ src, rule: RULE });

		expect(result.errorCount).toBe(count);
	});
});