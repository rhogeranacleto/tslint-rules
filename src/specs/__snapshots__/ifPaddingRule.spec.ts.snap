// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IfPaddingRule console.log(true);
		if(true){
			return true;
		} 1`] = `
"console.log(true);

		if(true){
			return true;
		}"
`;

exports[`IfPaddingRule console.log(true);
		if(true){
			return true;
		}else{
			return false;
		} 1`] = `
"console.log(true);

		if(true){
			return true;
		}else{
			return false;
		}"
`;

exports[`IfPaddingRule if(true){
			return 'so if';
		}
		else
		{
			return 'else if 2';
		} 1`] = `
"if(true){
			return 'so if';
		} else {
			return 'else if 2';
		}"
`;

exports[`IfPaddingRule if(true){
			return 'so if';
		}
		else
		if(false&&maria) {
			return 'else if';
		}
		else {
			return 's贸 else';
		} 1`] = `
"if(true){
			return 'so if';
		} else if(false&&maria) {
			return 'else if';
		} else {
			return 's贸 else';
		}"
`;

exports[`IfPaddingRule if(true){
			return 'so if';
		}
		else if(false&&maria) {
			return 'else if';
		} 1`] = `
"if(true){
			return 'so if';
		} else if(false&&maria) {
			return 'else if';
		}"
`;

exports[`IfPaddingRule if(true){
			return 'so if';
		} else
		{
			return 'else if';
		} 1`] = `
"if(true){
			return 'so if';
		} else {
			return 'else if';
		}"
`;

exports[`IfPaddingRule if(true){
			return 's贸 if';
		}
		else {
			return 'so else';
		} 1`] = `
"if(true){
			return 's贸 if';
		} else {
			return 'so else';
		}"
`;
